src/sinks/util/tcp.rs

```diff
@@ -108,249 +94,157 @@ impl TcpConnector {
         }
     }
 
-    fn connect(&self) -> BoxFuture<'static, Result<TcpOrTlsStream, TcpError>> {
-        let host = self.host.clone();
-        let port = self.port;
-        let resolver = self.resolver;
-        let tls = self.tls.clone();
+    fn fresh_backoff() -> ExponentialBackoff {
+        // TODO: make configurable
+        ExponentialBackoff::from_millis(2)
+            .factor(250)
+            .max_delay(Duration::from_secs(60))
+    }
 
-        async move {
-            let ip = resolver
-                .lookup_ip(host.clone())
-                .await
-                .context(DnsError)?
-                .next()
-                .ok_or(TcpError::NoAddresses)?;
+    async fn connect(&self) -> Result<MaybeTlsStream<TcpStream>, TcpError> {
+        let ip = self
+            .resolver
+            .lookup_ip(self.host.clone())
+            .await
+            .context(DnsError)?
+            .next()
+            .ok_or(TcpError::NoAddresses)?;
 
-            let addr = SocketAddr::new(ip, port);
-            let stream = tls.connect(host, addr).await.context(ConnectError)?;
-            Ok(FramedWrite::new(stream, BytesCodec::new()))
-        }
-        .boxed()
+        let addr = SocketAddr::new(ip, self.port);
+        self.tls
+            .connect(&self.host, &addr)
+            .await
+            .context(ConnectError)
     }
 
-    fn healthcheck(&self) -> BoxFuture<'static, crate::Result<()>> {
-        self.connect().map_ok(|_| ()).map_err(Into::into).boxed()
+    async fn connect_backoff(&self) -> MaybeTlsStream<TcpStream> {
+        let mut backoff = Self::fresh_backoff();
+        loop {
+            match self.connect().await {
+                Ok(socket) => {
+                    emit!(TcpSocketConnectionEstablished {
+                        peer_addr: socket.peer_addr().ok(),
+                    });
+                    return socket;
+                }
+                Err(error) => {
+                    emit!(TcpSocketConnectionFailed { error });
+                    delay_for(backoff.next().unwrap()).await;
+                }
+            }
+        }
     }
-}
 
-impl Into<TcpSink> for TcpConnector {
-    fn into(self) -> TcpSink {
-        TcpSink::new(self.host, self.port, self.resolver, self.tls)
+    async fn healthcheck(&self) -> crate::Result<()> {
+        self.connect().await.map(|_| ()).map_err(Into::into)
     }
 }
 
-pub struct TcpSink {
+struct TcpSink {
     connector: TcpConnector,
-    state: TcpSinkState,
-    backoff: ExponentialBackoff,
-    span: tracing::Span,
-}
-
-enum TcpSinkState {
-    Disconnected,
-    Connecting(Box<dyn Future<Item = TcpOrTlsStream, Error = TcpError> + Send>),
-    Connected(TcpOrTlsStream01, OpenTokenDyn),
-    Backoff(Box<dyn Future<Item = (), Error = ()> + Send>),
+    events_counter: Arc<EventsCounter>,
```

Comment:
Everything in these sinks run in a single thread, so I don't think we need to use mutex. I think we can get away with borrow checker issue by using Option<EventsCounter>, take and move it into `BytesSink`, then take it out of BytesSink when a send error happens. This will require calling `encode_event` in `start_send` of `BytesSink`, mentioned in another comment.

I have not tried this out so not sure if it actually work.

Reply:
I agree with the above comment.

I partly see what lead to this design, but there is no need for it. Since `send_all` of `futures` 0.3 doesn't consume the sink we can recover `EventsCounter` from it. With that there is no need for any mutex or atomic.  

And even if the above is wrong, `Mutex` is expensive to call for every event especially since there aren't other threads accessing this, so this complexity and performance cost are really not worth it.

Reply:
If we move `encode_event` function to `BytesSink` how we will able count events bytes in `EventsCounter`?

Reply:
`EventsCounter` could have method like `fn add_event(&self, byte_size: usize)` which `BytesSink` would call on successful encode of an event.